flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractFlinkResourceReconciler.java

```diff
@@ -384,6 +384,12 @@ private boolean shouldRollBack(
             return false;
         }
 
+        if (resource.getSpec().getJob() != null
+                && resource.getSpec().getJob().getUpgradeMode() == UpgradeMode.SAVEPOINT) {
+            // HA data is not available during rollback for savepoint upgrade mode
+            return true;
+        }
```

Comment:
This check is not fully correct. Just because we started from a savepoint if the jobmanager started succesfully HA metadata would be available even if the job could not recover after the upgrade. 

Instead of simply checkping for savepoint upgrade mode we also have to check `FlinkUtils.jmPodNeverStarted(ctx)` to be completely safe.

Reply:
Sorry but I'm not sure to see the issue if we just checkping for savepoint upgrade.
If the HA metadata has indeed been created by the jobmanger but the job has not recover we should not have any new checkpoint, am I wrong?
As in the rollback section we ensure that HA metadata is well deleted (in `cancelJob`) it should still be safe to 'not care' about this HA metadata, no?

Reply:
This is not that simple. The operator only periodically (lets say every 30sec) observes the job. It is possible that the job started, checkpointed and failed in that 30 sec period so we cannot simply ignore the HA metadata. If it's there we have to use it, we cannot delete it.

Reply:
Ok that was the thing I was missing

So we should indeed check that JM never start and ensure we rely on HA metadata if it exist otherwise rely on savepoint